File: backend/week1/day3-node-rest-api/src/controllers/productController.js
const { products } = require("../data/products");
const { successResponse } = require("../utils/apiResponse");
const AppError = require("../utils/appError");
const Product = require("../models/productModel");

// GET all products
exports.getAllProducts = async (req, res, next) => {
  try {
    const products = await Product.find();

    return successResponse(res, "All products fetched successfully", products);
  } catch (error) {
    return next(new AppError(error.message || "Failed to fetch products", 500));
  }
};

// GET product by ID
exports.getProductById = async (req, res, next) => {
  try {
    const product = await Product.findById(req.params.id);

    if (!product) {
      return next(new AppError("Product not found", 404));
    }

    return successResponse(res, "Product fetched successfully", product);
  } catch (error) {
    return next(new AppError("Invalid ID or failed to fetch product", 500));
  }
};

// CREATE product (POST)
exports.createProduct = async (req, res, next) => {
  try {
    const newProduct = await Product.create(req.body);

    return successResponse(res, "Product created successfully", newProduct);
  } catch (error) {
    return next(new AppError(error.message || "Failed to create product", 500));
  }
};

// UPDATE product (PUT)
exports.updateProduct = async (req, res, next) => {
  try {
    const updatedProduct = await Product.findByIdAndUpdate(
      req.params.id,
      req.body,
      {
        new: true, // return updated doc
        runValidators: true, // apply schema validation
      },
    );

    if (!updatedProduct) {
      return next(new AppError("Product not found", 404));
    }

    return successResponse(res, "Product updated successfully", updatedProduct);
  } catch (error) {
    return next(new AppError(error.message || "Failed to update product", 500));
  }
};

// PATCH
exports.patchProduct = async (req, res, next) => {
  try {
    const updatedProduct = await Product.findByIdAndUpdate(
      req.params.id,
      req.body,
      {
        new: true,
        runValidators: true,
      },
    );

    if (!updatedProduct) {
      return next(new AppError("Product not found", 404));
    }

    return successResponse(res, "Product updated partially", updatedProduct);
  } catch (error) {
    return next(new AppError(error.message || "Failed to patch product", 500));
  }
};
// DELETE
exports.deleteProduct = async (req, res, next) => {
  try {
    const deletedProduct = await Product.findByIdAndDelete(req.params.id);

    if (!deletedProduct) {
      return next(new AppError("Product not found", 404));
    }

    return successResponse(res, "Product deleted successfully", deletedProduct);
  } catch (error) {
    return next(new AppError(error.message || "Failed to delete product", 500));
  }
};

File: FAST APIs/E-Commerce/app/controllers/product_controller.py
from fastapi import FastAPI, APIRouter, HTTPException

from app.models.product_model import Product, ProductUpdate
from app.services.product_service import get_all_products, get_product_by_id, create_product, replace_product, patch_update, delete_product

router = APIRouter(
    prefix="/api/products",
    tags=["Products"]
)

# Get All products
@router.get("/")
def get_products():
    products = get_all_products()
    return products

# Get product by ID
@router.get("/{product_id}")
def get_product(product_id: int):
    product = get_product_by_id(product_id)
    if not product:
        raise HTTPException(status_code=404, detail="Product not found")
    
    return product

# Create new product
@router.post("/")
def add_product(product: Product):
    return create_product(product)

# Put product
@router.put("/{product_id}")
def update_product(product_id: int, product: Product):
    updated_product = replace_product(product_id, product)

    if not updated_product:
        raise HTTPException(status_code=404, detail="Product not found !")

    return updated_product

# Patch product
@router.patch("/{product_id}")
def update_partial_product(product_id: int, patch_data: ProductUpdate):
    patched_product = patch_update(product_id, patch_data)

    if not patched_product:
        raise HTTPException(status_code=404, detail="Product not found !")

    return patched_product

# Delete product
@router.delete("/{product_id}")
def delete_product_route(product_id: int):
    deleted = delete_product(product_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Product not found")
    return {"message": f"Product {product_id} deleted successfully"}